src/libcollections/string.rs

```diff
@@ -1131,10 +1147,12 @@ impl String {
         let len = self.len();
         assert!(idx <= len);
         assert!(self.is_char_boundary(idx));
-        let bits = ch.encode_utf8();
+        self.vec.reserve(4);
```

Comment:
I think you can avoid the `reserve` and intermediate buffers here w/ code similar to before and temporary buffers